discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -918,6 +931,85 @@ impl SourcePreferenceScope for SqliteStorage {
     }
 }
 
+#[derive(FromRow)]
+struct QueriedSourceReaction {
+    #[allow(dead_code)]
+    source: String,
+    #[allow(dead_code)]
+    weight: i32,
+    liked: bool,
+}
+
+#[async_trait]
+impl SourceReactionScope for SqliteStorage {
+    async fn fetch_source_reaction(&self, source: &str) -> Result<Option<bool>, Error> {
+        let mut tx = self.pool.begin().await?;
+
+        let reaction = sqlx::query_as::<_, QueriedSourceReaction>(
+            "SELECT source, weight, liked
+            FROM SourceReaction
+            WHERE source = ?;",
+        )
+        .bind(source)
+        .fetch_optional(&mut tx)
+        .await?;
+
+        tx.commit().await?;
+
+        Ok(reaction.map(|r| r.liked))
+    }
+
+    async fn store_source_reaction(&self, source: &str, liked: bool) -> Result<(), Error> {
+        let weight = if liked { 1 } else { -1 };
```

Comment:
what if we call `store_source_reaction` multiple times for the same source? we always reset the weight?

Reply:
right now it's not possible because first we check if we have a source with `fetch_source_reaction` and then later call `store_source_reaction` if the former returns `None`, but then maybe we should rename it to `create_source_reaction` (or add a line of documentation) to indicate that it should be only used for that and not for persisting source reaction (so creating and/or updating), wdyt?

Reply:
I agree the name was confusing, now changed as suggested! if you did call this multiple times for the same source, it should be a sql error (will check this when tests added).

Reply:
ah, true, because it's an `INSERT` operation, sorry, I'm not so fluent in SQL